dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java

```diff
@@ -255,10 +271,96 @@ private Grid getAggregatedDataValuesTableLayout(
 
     Map<String, Object> valueMap = AnalyticsUtils.getAggregatedDataValueMapping(grid);
 
+    AlternativeGridItems alternativeGridItems =
+        getAlternateGridItems(
+            grid, columnsDimensionItemsByDimension, rowsDimensionItemsByDimension, columns, rows);
+
+    alternativeGridItems =
+        AlternativeGridItems.of(
+            reorderItems(alternativeGridItems.getRows()),
+            reorderItems(alternativeGridItems.getColumns()));
+
     return visualization.getGrid(
         new ListGrid(grid.getMetaData(), grid.getInternalMetaData()),
         valueMap,
+        alternativeGridItems,
         params.getDisplayProperty(),
         false);
   }
+
+  private List<List<DimensionalItemObject>> reorderItems(
+      List<List<DimensionalItemObject>> alternateItems) {
+    return alternateItems.stream().sorted(this::compareItems).collect(Collectors.toList());
+  }
+
+  private int compareItems(List<DimensionalItemObject> dio, List<DimensionalItemObject> other) {
+    return dio.get(0).getDisplayName().compareTo(other.get(0).getDisplayName());
+  }
+
+  /**
+   * Returns alternative grid rows and columns based on the given grid and dimension items.
+   * Alternative grid items are used improve the performance of the grid rendering when the
+   * combination of dimension items is large. The alternative grid items are a list of lists of
+   * dimension items where each list represents a possible combination of dimension items that are
+   * used to render the grid.
+   *
+   * @param grid the grid
+   * @param columnsDimensionItemsByDimension a map of dimension items by dimension for columns
+   * @param rowsDimensionItemsByDimension a map of dimension items by dimension for rows
+   * @param columnDimensionIds the column dimension ids
+   * @param rowDimensionIds the row dimension ids
+   * @return the alternative grid items
+   */
+  public static AlternativeGridItems getAlternateGridItems(
```

Comment:
This method can be simplified to work only with a Map<String, List<DimensionalItemObject>> and a List List<String>.

So, you can call it `getDimensionItemsFromMap`, or something similar.

Then the consumers will do:

`Map<String, List<DimensionalItemObject>> columnsDimensionItemsByDimension = getDimensionItemsFromMap(...);`

`Map<String, List<DimensionalItemObject>> rowsDimensionItemsByDimension = getDimensionItemsFromMap(...);`

And we eliminate the class `AlternativeGridItems`